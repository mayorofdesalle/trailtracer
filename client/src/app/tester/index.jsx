import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
