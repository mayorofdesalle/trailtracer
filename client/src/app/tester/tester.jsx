import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };