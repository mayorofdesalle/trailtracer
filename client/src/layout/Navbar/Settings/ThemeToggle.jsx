import { useEffect, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import styled, { useTheme } from 'styled-components';

import { toggle } from '@features/themeSlice';
import Container from '@components/ui/Container';
import Button from '@components/ui/Button';
import Text from '@components/ui/Text';
import Icon from '@components/ui/Icon';

const ThemeToggleContainer = styled(Container)`
    height: fit-content;
    padding: 0;
    justify-content: space-between;
`;

const ToggleBackground = styled(Container)`
    width: clamp(6rem, min(12dvw, 12dvh), 24rem);
    height: clamp(1rem, min(2dvw, 2dvh), 2rem);
    background-color: ${({ theme }) => theme.colors.primary};

    & > ${Button} {
        position: absolute;
        padding: clamp(0.5rem, min(1dvw, 1dvh), 2rem);
        transform: translateX(${({ $preference }) => $preference ? '-100%' : '100%'});
        background-color: ${({ theme }) => theme.colors.background};
        border: 2px solid ${({ theme }) => theme.colors.primaryTransparent};

        &:hover, &:active {
            border-color: ${({ theme }) => theme.colors.backgroundTransparent};
        }
    }
`;

/**
 * ThemeToggle
 * @description
 * Toggle to switch between light and dark themes.
 **/
const ThemeToggle = () => {
    const theme = useTheme();
    const { t } = useTranslation();
    const dispatch = useDispatch();
    const preference = useSelector((state) => state.theme.preference);

    useEffect(() => {
        document.querySelector('meta[name="theme-color"]').setAttribute('content', preference ? '#000000' : '#FCFFFD');
    }, [preference]);

    const onClick = useCallback(() => dispatch(toggle()), [dispatch]);

    return (
        <ThemeToggleContainer>
            <Text $heading $ratio={0.5}>{t('forms.theme')}</Text>
            <ToggleBackground $preference={preference}>
                <Button onClick={onClick}>
                    <Icon name={preference ? 'moon-fill' : 'sun-fill'} color={theme.colors.primary} />
                </Button>
            </ToggleBackground>
        </ThemeToggleContainer>
    );
};

export default ThemeToggle;