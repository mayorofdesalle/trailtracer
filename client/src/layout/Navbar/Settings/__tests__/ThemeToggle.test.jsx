import { describe, expect, it, render, screen } from '@app/tester';
import userEvent from '@testing-library/user-event';

import ThemeToggle from '../ThemeToggle';

describe('ThemeToggle', () => {

    it('should render correctly', () => {
        render(<ThemeToggle data-testid='theme-toggle' />);
        const toggle = screen.getByTestId('theme-toggle');
        expect(toggle).toBeInTheDocument();
        expect(toggle).toHaveTextContent(/theme/i);

        const icon = screen.getByRole('img');
        expect(icon).toBeInTheDocument();

        const button = screen.getByRole('switch');
        expect(button).toBeInTheDocument();
    });

    it('should toggle theme', async () => {
        const user = userEvent.setup();

        render(<ThemeToggle />);
        const button = screen.getByRole('switch');
        expect(button).toBeInTheDocument();
        expect(button).toHaveAttribute('aria-checked', 'false');

        await user.click(button);
        expect(button).toHaveAttribute('aria-checked', 'true');

        await user.click(button);
        expect(button).toHaveAttribute('aria-checked', 'false');
    });

    it('should match snapshot', () => {
        const { asFragment } = render(<ThemeToggle />);
        expect(asFragment()).toMatchSnapshot();
    });

});