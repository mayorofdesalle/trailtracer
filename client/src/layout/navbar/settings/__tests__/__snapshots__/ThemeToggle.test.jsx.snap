// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeToggle > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-guDLey sc-kFCroH kHwwsn leamf"
  >
    <div
      class="sc-gLLuof dcSHkq"
    >
      forms.theme
    </div>
    <div
      class="sc-guDLey sc-irLvIq kHwwsn cdxcFH"
    >
      <button
        aria-checked="false"
        class="sc-beySPh zbkUW"
        role="switch"
        type="button"
      >
        <svg
          class="sc-eDLKkx fuXsHX"
          fill="#435049"
          role="img"
        >
          <title>
            sun-fill
          </title>
          <use
            href="http://localhost:3000/src/assets/icons.svg#sun-fill"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
